### PR TITLE
bpo-29446: Improve tkinter 'import *' situation

### PR DESCRIPTION
**Pull request by flavianh**
A pull request for [this issue](https://bugs.python.org/issue29446).

@ericvsmith recommended using `__all__ = [name for name, obj in globals().items() if not name.startswith('_') and not isinstance(obj, types.ModuleType) and name not in {'wantobjects'}]`.

Also added __all__ in submodules of tkinter.

Side note: Interestingly, `idlelib/editor.py` was using `sys` and `re` from `tkinter` instead of importing them. There may be more modules that use "from tkinter import *"

https://bugs.python.org/issue29446

### CODE REVIEW

File: Lib/tkinter/__init__.py [hunk 1]
```diff
@@ -30,6 +30,85 @@
...
+    "messagebox",
+    "scrolledtext",
+    "simpledialog",
+    "tix",
+    "ttk",
+]
+
+__all__internals_used_in_submodules = [
+    "_cnfmerge",      # Used in tkinter.dialog
+    "_default_root",  # Used in tkinter.simpledialog
```
Comment (serhiy-storchaka) [inline Lib/tkinter/__init__.py]: `_default_root` should **never** be imported, neither implicitly, nor implicitly.

`_cnfmerge` should not be imported implicitly since it is an internal function.

File: Lib/tkinter/__init__.py [hunk 2]
```diff
@@ -30,6 +30,85 @@
...
+]
+
+__all__ = __all__base + __all__constants + __all__modules + __all__internals_used_in_submodules
```
Comment (serhiy-storchaka) [inline Lib/tkinter/__init__.py]: Why introduce all these lists? Why not use just `__all__`?

File: Lib/tkinter/__init__.py [hunk 3]
```diff
@@ -30,6 +30,85 @@
...
+    "image_names",
+    "image_types",
+    "Spinbox",
+    "LabelFrame",
+    "PanedWindow",
+]
+
+__all__constants = [item for item in dir(tkinter.constants) if not item.startswith("__")]
+
+__all__modules = [
```
Comment (serhiy-storchaka) [inline Lib/tkinter/__init__.py]: Including submodule names will make `from tkinter import *` importing all submodules, even if they are not used. I would exclude submodule names from `__all__`.

File: Lib/tkinter/__init__.py [hunk 4]
```diff
@@ -30,6 +30,85 @@
 tk.mainloop()
 """
 
+import tkinter.constants
+
+__all__base = [
```
Comment (serhiy-storchaka) [inline Lib/tkinter/__init__.py]: This list is too long. I prefer the approach proposed by @ericvsmith.

File: Lib/tkinter/constants.py [hunk 5]
```diff
@@ -1,110 +1,110 @@
 # Symbolic constants for Tk
 
 # Booleans
-NO=FALSE=OFF=0
-YES=TRUE=ON=1
+NO = FALSE = OFF = 0
```
Comment (serhiy-storchaka) [inline Lib/tkinter/constants.py]: This is an unrelated change.

File: Lib/tkinter/dialog.py [hunk 6]
```diff
@@ -1,7 +1,12 @@
 # dialog.py -- Tkinter interface to the tk_dialog script.
 
-from tkinter import *
-from tkinter import _cnfmerge
+from tkinter import (
```
Comment (serhiy-storchaka) [inline Lib/tkinter/dialog.py]: Would be nice to add explicit `__all__` in submodules.

File: Lib/tkinter/filedialog.py [hunk 7]
```diff
@@ -12,13 +12,30 @@
...
+    Button,
+    Entry,
+    Frame,
+    Listbox,
+    Scrollbar,
+    Tk,
+    Toplevel,
+    commondialog,
+)
+from tkinter.constants import (
```
Comment (serhiy-storchaka) [inline Lib/tkinter/filedialog.py]: Could not this be written in one line?

File: Lib/tkinter/tix.py [hunk 8]
```diff
@@ -24,10 +24,6 @@
 
 import os
 import tkinter
-from tkinter import *
```
Comment (serhiy-storchaka) [inline Lib/tkinter/tix.py]: I would not touch `tix.py`. It is deprecated and may be removed in 3.9.

**Review by terryjreedy — CHANGES_REQUESTED** [review]
Flavian, thank you for making a start on this;  As for changes, I pretty much agree with Serhiy.
1  Use Eric's suggestion for things defined in `.__init__` and exclude submodules.
2. Make lists horizontal, with multiple items per line (this is pretty standard).
3. Put PEP 8 style changes in another PR, or even another issue;  but only after asking Serhiy if he wants them, or at least would merge them.
4. I never wrote about submodules, but `__all__` for each, except Tix (leave alone), would be nice.

File: Misc/NEWS.d/next/Library/2019-07-19-16-06-48.bpo-29446.iXGuoi.rst [hunk 9]
```diff
@@ -0,0 +1 @@
+Make `from tkinter import *` import only the expected objects.
```
Comment (terryjreedy) [inline Misc/NEWS.d/next/Library/2019-07-19-16-06-48.bpo-29446.iXGuoi.rst:1]:
```suggestion
Make `from tkinter import *` import only the objects it should.
```
Serhiy can decide if this is an improvement.

**Comment by terryjreedy** [comment]
With respect to the list in your initial comment: The 'Var's are absolutely essential.  Without checking idlelib, I presume the other non-underscore names you list there are used also.

File: Lib/tkinter/__init__.py (hunk 1 above)
Comment (flavianh) [reply Lib/tkinter/__init__.py]: `_default_root` is [already imported in tkinter.simpledialog](https://github.com/python/cpython/blob/7a131a8fcd0bd5db2956223dc6cf0cd94a07824f/Lib/tkinter/simpledialog.py#L278-L279) and `_cnfmerge` is [used in dialog.py](https://github.com/python/cpython/blob/7a131a8fcd0bd5db2956223dc6cf0cd94a07824f/Lib/tkinter/dialog.py#L16)

For `_cnfmerge` the change is straightforward: I just have to omit it in `__all__` and the import in dialog.py will work just fine

For `_default_root` I don't know how to remove it from `simpledialog.py` without affecting the behaviour of `_QueryDialog`. I can't leave `parent` as None if no parent is passed because then the line [`if parent.winfo_viewable():`](https://github.com/python/cpython/blob/7a131a8fcd0bd5db2956223dc6cf0cd94a07824f/Lib/tkinter/simpledialog.py#L153) will fail.

Same in `font.py`.

I updated the PR to not export `_cnfmerge` and `_default_root` in `__all__`, but could not prevent `_default_root` from being used elsewhere.

File: Lib/tkinter/__init__.py (hunk 2 above)
Comment (flavianh) [reply Lib/tkinter/__init__.py]: I'll use Eric's solution so this won't be an issue

File: Lib/tkinter/__init__.py (hunk 3 above)
Comment (flavianh) [reply Lib/tkinter/__init__.py]: OK makes sense

File: Lib/tkinter/constants.py (hunk 5 above)
Comment (flavianh) [reply Lib/tkinter/constants.py]: True, I dismissed those changes. Can I open a new PR to add this change?

File: Lib/tkinter/dialog.py (hunk 6 above)
Comment (flavianh) [reply Lib/tkinter/dialog.py]: I'll gladly include that

File: Lib/tkinter/filedialog.py (hunk 7 above)
Comment (flavianh) [reply Lib/tkinter/filedialog.py]: I was unsure about the code style here. What's the standard?
I reworked the imports though, so there might not be any problem any more

File: Lib/tkinter/tix.py (hunk 8 above)
Comment (flavianh) [reply Lib/tkinter/tix.py]: Great news, it seemed like a mess :)

**Comment by flavianh** [comment]
I have made the requested changes; please review again

File: Lib/tkinter/__init__.py [hunk 10]
```diff
@@ -34,11 +34,13 @@
 import sys
 
 import _tkinter # If this fails your Python may not be configured for Tk
+
+import types
```
Comment (serhiy-storchaka) [inline Lib/tkinter/__init__.py:35]: Move it up, to `enum` and `sys`.

File: Lib/tkinter/colorchooser.py [hunk 11]
```diff
@@ -19,8 +19,9 @@
 # - title: dialog title
 #
 
-from tkinter.commondialog import Dialog
+__all__ = ["Chooser", "askcolor"]
 
+from tkinter.commondialog import Dialog
```
Comment (serhiy-storchaka) [inline Lib/tkinter/colorchooser.py]: PEP 8: Imports are always put at the top of the file, just after any module comments and docstrings, and before module globals and constants.

File: Lib/tkinter/commondialog.py [hunk 12]
```diff
@@ -8,15 +8,12 @@
 # written by Fredrik Lundh, May 1997
 #
 
-from tkinter import *
```
Comment (serhiy-storchaka) [inline Lib/tkinter/commondialog.py:11]: How can it work without any import?

File: Lib/tkinter/dnd.py [hunk 13]
```diff
@@ -99,8 +99,9 @@
 
 """
 
+__all__ = ["dnd_start", "DndHandler"]
 
-import tkinter
+from tkinter import *
```
Comment (serhiy-storchaka) [inline Lib/tkinter/dnd.py]: Why is this change?

File: Lib/tkinter/filedialog.py [hunk 14]
```diff
@@ -11,13 +11,15 @@
...
 These interfaces were written by Fredrik Lundh, May 1997.
 """
+__all__ = ["FileDialog", "LoadFileDialog", "SaveFileDialog", "Open", "SaveAs",
+           "Directory", "askopenfilename", "asksaveasfilename", "askopenfilenames",
+           "askopenfile", "askopenfiles", "asksaveasfile", "askdirectory"]
 
+import fnmatch
+import os
 from tkinter import *
+from tkinter.commondialog import Dialog as CommonDialog
```
Comment (serhiy-storchaka) [inline Lib/tkinter/filedialog.py]: Why is this change?

File: Lib/tkinter/scrolledtext.py [hunk 15]
```diff
@@ -13,8 +13,7 @@
 
 __all__ = ['ScrolledText']
 
-from tkinter import Frame, Text, Scrollbar, Pack, Grid, Place
-from tkinter.constants import RIGHT, LEFT, Y, BOTH
+from tkinter import *
```
Comment (serhiy-storchaka) [inline Lib/tkinter/scrolledtext.py]: It is better to import only these names which are needed.

File: Lib/tkinter/simpledialog.py [hunk 16]
```diff
@@ -25,8 +25,7 @@
 
 from tkinter import *
 from tkinter import messagebox
-
-import tkinter # used at _QueryDialog for tkinter._default_root
+from tkinter import _default_root
```
Comment (serhiy-storchaka) [inline Lib/tkinter/simpledialog.py]: *Never* import `_default_root` as its value usually is changed after importing.

File: Lib/tkinter/test/test_tkinter/test_misc.py [hunk 17]
```diff
@@ -7,6 +7,19 @@
 
 class MiscTest(AbstractTkTest, unittest.TestCase):
 
+    def test_all(self):
+        self.assertIn("Widget", tkinter.__all__)
+        # Check that variables from tkinter.constants are also in tkinter.__all__
+        self.assertIn("CASCADE", tkinter.__all__)
+        self.assertIsNotNone(tkinter.CASCADE)
+        # Check that sys or re are not in tkiner.__all__
```
Comment (serhiy-storchaka) [inline Lib/tkinter/test/test_tkinter/test_misc.py]: Please check also "constants".

File: Lib/tkinter/test/test_tkinter/test_misc.py [hunk 18]
```diff
@@ -7,6 +7,19 @@
...
+        self.assertNotIn("re", tkinter.__all__)
+        self.assertNotIn("sys", tkinter.__all__)
+        # Check that an underscored functions is not in tkiner.__all__
+        self.assertNotIn("_tkerror", tkinter.__all__)
+        # Check that wantobjects functions is not in tkiner.__all__
```
Comment (serhiy-storchaka) [inline Lib/tkinter/test/test_tkinter/test_misc.py]: `wantobjects` is not a function.

File: Misc/ACKS [hunk 19]
```diff
@@ -1874,3 +1874,4 @@ Edison Abahurire
 Geoff Shannon
 Batuhan Taskaya
 Aleksandr Balezin
+Flavian Hautbois
```
Comment (serhiy-storchaka) [inline Misc/ACKS]: Place your name at appropriate place according to dictionary order (between "Hauser" and "Hawkes").

File: Lib/tkinter/__init__.py (hunk 1 above)
Comment (serhiy-storchaka) [reply Lib/tkinter/__init__.py]: No, it is not imported, but is accessed as an attribute of `tkinter`.

The difference is that its value is changed after importing. When you import the submodule, it is None. After creating a root widget it keeps a reference to it, which can be used as a default parent in `_QueryDialog`.

Comment (flavianh) [reply Lib/tkinter/__init__.py]: Thanks for the clarification!

File: Lib/tkinter/commondialog.py (hunk 12 above)
Comment (flavianh) [reply Lib/tkinter/commondialog.py:11]: My bad

File: Lib/tkinter/dnd.py (hunk 13 above)
Comment (flavianh) [reply Lib/tkinter/dnd.py]: I wanted to make it more similar to the other files. But I get your point: keep changes minimal, this is indeed just for style

File: Lib/tkinter/font.py [hunk 20]
```diff
@@ -3,11 +3,12 @@
 # written by Fredrik Lundh, February 1998
 #
 
-__version__ = "0.9"
-
 import itertools
 import tkinter
 
+__version__ = "0.9"
```
Comment (flavianh) [inline Lib/tkinter/font.py:9]: Had to move `__version__` here to abide to PEP8 also

**Comment by flavianh** [comment]
@serhiy-storchaka All you comments should now be addressed.

**Comment by flavianh** [comment]
@terryjreedy @serhiy-storchaka Anything blocking?

File: Lib/idlelib/editor.py [hunk 21]
```diff
@@ -2,7 +2,9 @@
 import importlib.util
 import os
 import platform
+import re
 import string
+import sys
```
Comment (terryjreedy) [inline Lib/idlelib/editor.py]: I am adding these imports to editor.py separately, and backport to 3.8,3.7.  I will then remove them here.  This issue should be backported to 3.8 but not 3.7.

Done in #14919

**Review by terryjreedy — COMMENTED** [review]
Serhiy is responsible for final details of tkinter/* changes.

File: Lib/idlelib/editor.py (hunk 21 above)
Comment (flavianh) [reply Lib/idlelib/editor.py]: OK, do I need to do anything?

File: Lib/tkinter/filedialog.py [hunk 22]
```diff
@@ -11,14 +11,16 @@
 directory dialogue available in Tk 8.3 and newer.
 These interfaces were written by Fredrik Lundh, May 1997.
 """
+__all__ = ["FileDialog", "LoadFileDialog", "SaveFileDialog", "Open", "SaveAs",
+           "Directory", "askopenfilename", "asksaveasfilename", "askopenfilenames",
+           "askopenfile", "askopenfiles", "asksaveasfile", "askdirectory"]
 
+import fnmatch
+import os
 from tkinter import *
```
Comment (terryjreedy) [inline Lib/tkinter/filedialog.py]: I would prefer that this be replace by explicit imports, as done in dialog.py and commondialog.py

File: Lib/idlelib/editor.py (hunk 21 above)
Comment (terryjreedy) [reply Lib/idlelib/editor.py]: No, separate PR merged and backports in progress.

**Comment by flavianh** [comment]
@terryjreedy I had to remove your commit "Move editor.py changes to another PR for full backport" because it was undoing your own changes when I rebased on master ;)

**Comment by terryjreedy** [comment]
Serhiy, as near as I can tell, this is ready to go.  If you agree, it would be nice to get it into a3, due next Monday.

File: Lib/tkinter/commondialog.py [hunk 23]
```diff
@@ -8,15 +8,15 @@
...
+from tkinter import Frame
```
Comment (serhiy-storchaka) [inline Lib/tkinter/commondialog.py:13]: Maybe add `__all__`?

File: Lib/tkinter/font.py [hunk 24]
```diff
@@ -3,11 +3,12 @@
...
+__all__ = ["NORMAL", "ROMAN", "BOLD", "ITALIC", "nametofont", "Font", "families",
```
Comment (serhiy-storchaka) [inline Lib/tkinter/font.py]: I would break a line after `"ITALIC"`.

File: Lib/tkinter/messagebox.py [hunk 25]
```diff
@@ -24,6 +24,9 @@
 
 from tkinter.commondialog import Dialog
 
+__all__ = ["showinfo", "showwarning", "showerror", "askquestion", "askokcancel",
```
Comment (serhiy-storchaka) [inline Lib/tkinter/messagebox.py]: I suggest to break a line after `"showerror"`.

File: Lib/tkinter/filedialog.py [hunk 26]
```diff
@@ -11,14 +11,19 @@
 directory dialogue available in Tk 8.3 and newer.
 These interfaces were written by Fredrik Lundh, May 1997.
 """
+__all__ = ["FileDialog", "LoadFileDialog", "SaveFileDialog", "Open", "SaveAs",
+           "Directory", "askopenfilename", "asksaveasfilename", "askopenfilenames",
```
Comment (serhiy-storchaka) [inline Lib/tkinter/filedialog.py]: This line is longer then PEP 8 limit. I would format the list like:
```
__all__ = ["FileDialog", "LoadFileDialog", "SaveFileDialog",
           "Open", "SaveAs", "Directory",
           "askopenfilename", "asksaveasfilename", "askopenfilenames",
           "askopenfile", "askopenfiles", "asksaveasfile", "askdirectory"]
```

File: Lib/tkinter/font.py (hunk 24 above)
Comment (flavianh) [reply Lib/tkinter/font.py]: Sure thing. What is your rule for line breaks? I'd like to automate it in my IDE

File: Lib/tkinter/filedialog.py (hunk 26 above)
Comment (flavianh) [reply Lib/tkinter/filedialog.py]: Added the 79-character line length to my IDE's configuration for cPython

**Comment by flavianh** [comment]
@serhiy-storchaka Changes done ;)

**Review by serhiy-storchaka — APPROVED** [review]

File: Lib/tkinter/font.py (hunk 24 above)
Comment (serhiy-storchaka) [reply Lib/tkinter/font.py]: It is hard to automate. Semantically the first four names in this list are names of constants.

**Comment by flavianh** [comment]
@terryjreedy up to you :)

**Comment by terryjreedy** [comment]
It doesn't really matter which of us merges.  I presume that this should not be backported to 3.8 because b3 is too late for new features, even __all__.  There must be other code that will break with this change, just like IDLE did.
